main.go

```diff
@@ -80,214 +76,232 @@ func main() {
 	glog.Info("Terminated")
 }
 
-// Handles updates for all services and dispatch them between specific per service handlers.
-func handleServices(c *configuration, cloud cloud.Cloud, updates <-chan *registry.ServiceUpdate, done chan struct{}) {
+// Dispatches updates for all services between per service channels.
+func dispatchUpdates(c *configuration, cloud cloud.Cloud, updates <-chan *registry.ServiceUpdate, done chan struct{}) {
 	var wg sync.WaitGroup
-	handlers := make(map[string]chan *registry.ServiceUpdate)
+	serviceUpdatesMap := make(map[string]chan *registry.ServiceUpdate)
 
 	for {
 		select {
 		case update := <-updates:
-			glog.Infof("Handling %s service update", update.ServiceName)
-			// is there a handler for updated service?
-			if handler, ok := handlers[update.ServiceName]; !ok {
+			glog.Infof("Got %q service update", update.ServiceName)
+			// is there a channel for service updates?
+			if serviceUpdates, ok := serviceUpdatesMap[update.ServiceName]; !ok {
 				// no handler
-				handler = make(chan *registry.ServiceUpdate)
-				handlers[update.ServiceName] = handler
+				serviceUpdates = make(chan *registry.ServiceUpdate)
+				serviceUpdatesMap[update.ServiceName] = serviceUpdates
 				// start handler in its own goroutine
 				wg.Add(1)
-				glog.Infof("Initializing a handler for %s service", update.ServiceName)
-				go handleService(c, cloud, update.Tag, handler, &wg, done)
+				glog.Infof("Initializing a handler for %q service updates...", update.ServiceName)
+
+				tagConfig, err := c.GetTagConfiguration(update.Tag)
+				if err != nil {
+					glog.Error(err)
+					continue
+				}
+
+				go handleServiceUpdates(c.Cloud.URLMap, cloud, tagConfig, serviceUpdates, &wg, done)
```

Comment:
You are spreading the control logic of wait group by passing it by pointer to the goroutine. A better approach would be to have something like this — 

```
go func(){
  handleServiceUpdates(c.Cloud.URLMap, cloud, tagConfig, serviceUpdates, done)
  wg.Done()
}()
```

Then you are not passing the wait group outside of the scope it was initially defined.